File: projections/blazor/ui/hlp.ui.layout-runtime/LayoutRuntimeTypes.cs
using Harborline.Foundation.RuleAuthoring;
using Harborline.UIAdapters.Blazor.Components.RuleAuthoring;

namespace Harborline.UIAdapters.Blazor.Components.Layout;

public sealed record LayoutRuntimeDiagnostic(string Code, string Pointer);
public sealed record LayoutRuntimeBlock(string Id, string Kind, int Depth, string? Zone = null);
/// <summary>DES-0052 C1, layout-eng-26: the authority the platform returned with the surface. The lane derives nothing from it.</summary>
public sealed record LayoutRuntimeAuthority(bool CanSubmit);
// Authority: absent, the surface is read-only (no authority, no submit).
/// <summary>DES-0056 execution-runtime-ck-4: one execution-trace step, verbatim.</summary>
public sealed record LayoutExecutionTraceStep(int Ordinal, string Phase, string Status);
/// <summary>DES-0056 execution-runtime-ck-3, as the host's authorized read returned it; a null <paramref name="AccessDecisionId"/> is no recorded decision.</summary>
public sealed record LayoutRunReceipt(string RunId, string Status, string? AccessDecisionId, IReadOnlyList<LayoutExecutionTraceStep> Trace);
/// <summary>One of Access's four stages, as stored.</summary>
public sealed record LayoutAccessTraceStage(int Ordinal, string Stage, IReadOnlyList<string> Facts);
/// <summary>
/// What following the Access decision link yielded, classified by the platform (LayoutExecutionObservation):
/// <c>absent</c>, <c>missing</c>, <c>forbidden</c>, <c>malformed</c> or <c>valid</c>. The lane derives nothing.
/// </summary>
public sealed record LayoutAccessTrace(string Evidence, int? Version, IReadOnlyList<LayoutAccessTraceStage> Stages, string? DecidingGrant);
public sealed record LayoutRuntimePlan(
    string DefinitionId,
    string DefinitionVersionId,
    string Medium,
    IReadOnlyList<LayoutRuntimeBlock> Flow,
    IReadOnlyList<LayoutRuntimeBlock> StaticRegions,
    IReadOnlyList<LayoutRuntimeDiagnostic>? Diagnostics = null,
    LayoutRuntimeAuthority? Authority = null);

/// <summary>
/// One authored binding: the kind (record_field, query, measure, template or static) and the
/// name it resolves by. An empty <paramref name="Name"/> is the `needs a binding` state — the
/// block is preserved and rebound one at a time rather than deleted (layout-auth-31).
/// </summary>
public sealed record LayoutAuthoringBinding(string Kind, string Name);

/// <summary>A Rules named predicate's exact pin (rules-ck-22): nothing floats.</summary>
public sealed record LayoutAuthoringPredicatePin(string Name, string Version, string Digest);

/// <summary>
/// layout-ck-29: a guard holds exactly one of a Rules expression, compiled at publish, or a named
/// predicate by exact pin, resolved from the definition's pinned closure. Both fail closed.
/// </summary>
public sealed record LayoutAuthoringShowWhen(string? Expression = null, LayoutAuthoringPredicatePin? Predicate = null);

/// <summary>A named predicate a show_when guard may cite (layout-ck-29).</summary>
public sealed record LayoutAuthoringPredicate(string Label, LayoutAuthoringPredicatePin Pin);

public sealed record LayoutAuthoringBlock(
    string Id,
    string Kind,
    LayoutAuthoringBinding? Binding = null,
    string? ParentId = null,
    string? Zone = null,
    string? Intent = null,
    string? Width = null,
    string? Height = null,
    string? AlignSelf = null,
    string? StaticRegion = null,
    string? WidgetId = null,
    bool BreakBefore = false,
    bool AvoidPageBreak = false,
    // layout-auth-18: the block repeats its children once per row of its collection binding.
    bool Repeating = false,
    // The flow this block arranges its children in; admission requires one on a repeating block or any parent (T-724 ruling 41).
    string? Container = null,
    // layout-auth-19: the declared Records relationship this block observes; only the key is stored.
    string? RelatedRelationship = null,
    // layout-ck-29, layout-auth-20: the block's guard, holding exactly one form; absent, the block always shows.
    LayoutAuthoringShowWhen? ShowWhen = null,
    // The guided expression ShowWhen was lowered from; absent when the guard was written as raw text (T-724 ruling 39).
    FormulaExpr? ShowWhenGuide = null,
    // The capture properties a capture block narrows with (layout-ck-30).
    LayoutAuthoringCapture? Capture = null,
    // layout-auth-33: the selection this block opens with. Authored; the live selection is never stored.
    string? DefaultSelection = null,
    // layout-auth-34: the ids of the other blocks on this surface this block's selection filters.
    IReadOnlyList<string>? FilterTargets = null);

/// <summary>
/// What a capture block narrows (layout-ck-30): it may add a requirement and name registered
/// validation rules (layout-auth-21), never remove what Records declared.
/// </summary>
public sealed record LayoutAuthoringCapture(
    bool Required = false,
    IReadOnlyList<string>? ValidationRules = null,
    // layout-auth-22: the prompt this surface shows for the field, in its own context only.
    string? PromptOverride = null,
    // layout-bound-3: the registered field control this capture field uses; absent is the runtime's choice.
    string? Control = null);
public sealed record LayoutAuthoringPageRun(string Id, string PageLayoutId, string PageMasterId);
public sealed record LayoutAuthoringDraft(
    string Name,
    string Medium,
    string? CollapseBelow,
    IReadOnlyList<LayoutAuthoringBlock> Blocks,
    string? DefaultIntent = null,
    string? ContainerFlow = null,
    int? Gap = null,
    string? Density = null,
    IReadOnlyList<LayoutAuthoringPageRun>? PageRuns = null,
    // layout-auth-35: the released surfaces a reader may drill through to from this one.
    IReadOnlyList<string>? DrillThroughTargets = null)
{
    public static LayoutAuthoringDraft Empty { get; } = new("", "screen", null, []);
}
public sealed record LayoutAuthoringOption(string Id, string Label);
public sealed record LayoutAuthoringCatalogue(
    IReadOnlyList<LayoutAuthoringOption> BlockKinds,
    IReadOnlyList<string> Zones,
    IReadOnlyList<LayoutAuthoringOption>? PageLayouts = null,
    IReadOnlyList<LayoutAuthoringOption>? PageMasters = null,
    IReadOnlyList<string>? StaticRegions = null,
    IReadOnlyList<LayoutAuthoringOption>? HelmWidgets = null,
    // Bindables: the names offered per binding kind; `static` is authored on the block.
    IReadOnlyDictionary<string, IReadOnlyList<LayoutAuthoringOption>>? Bindables = null,
    // Relationships: the Records relationships declared on the surface's record type, by key (layout-auth-19).
    IReadOnlyList<LayoutAuthoringOption>? Relationships = null,
    // RequiredFields: the record fields Records declares required; a capture block cannot drop them (layout-auth-21).
    IReadOnlyList<string>? RequiredFields = null,
    // ValidationRules: the registered validation rules a capture block may name (layout-auth-21, layout-bound-8).
    IReadOnlyList<LayoutAuthoringOption>? ValidationRules = null,
    // DrillTargets: the released surfaces a drill-through may name (layout-auth-35).
    IReadOnlyList<LayoutAuthoringOption>? DrillTargets = null,
    // FieldControls: the field controls the host registers for capture fields (layout-bound-3).
    IReadOnlyList<LayoutAuthoringOption>? FieldControls = null,
    // ValueDomainFields: the record fields whose value domain picks their editor; they take no authored control (layout-bound-10).
    IReadOnlyList<string>? ValueDomainFields = null,
    // GuardReferences: the references a show_when guard may read, offered by the guided expression editor (layout-auth-20).
    IReadOnlyList<RulesPaletteItem>? GuardReferences = null,
    // Predicates: the named predicates a show_when guard may cite, each by its exact pin (layout-ck-29).
    IReadOnlyList<LayoutAuthoringPredicate>? Predicates = null);

File: projections/react/ui/hlp.ui.layout-runtime/src/LayoutRuntime.types.ts
import type { RuleDefinitionExpression, RuleDefinitionValueType } from '@harborline-software/rule-authoring'

export type LayoutMedium = 'screen' | 'page'
export type LayoutContainerToken = 'sm' | 'md' | 'lg'
export type LayoutIntent = 'capture' | 'observe' | 'issue'
export type LayoutContainerFlow = 'stack' | 'flow' | 'areas'
export type LayoutAlignment = 'start' | 'center' | 'end' | 'stretch'
export type LayoutSizing = 'hug' | 'fill' | 'fixed'

export interface LayoutRuntimeDiagnostic { readonly code: string; readonly pointer: string }
export interface LayoutRuntimeBlock { readonly id: string; readonly kind: string; readonly depth: number; readonly zone?: string }
/** DES-0052 C1, layout-eng-26: the authority the platform returned with the surface. The lane derives nothing from it. */
export interface LayoutRuntimeAuthority { readonly canSubmit: boolean }
export interface LayoutRuntimePlan {
  readonly definitionId: string
  readonly definitionVersionId: string
  readonly medium: LayoutMedium
  readonly flow: readonly LayoutRuntimeBlock[]
  readonly staticRegions: readonly LayoutRuntimeBlock[]
  readonly diagnostics?: readonly LayoutRuntimeDiagnostic[]
  /** Absent, the surface is read-only: no authority, no submit. */
  readonly authority?: LayoutRuntimeAuthority
}
/** `onSubmit` hands the submit to the host, which re-asks the platform's gate before any write. */
export interface LayoutRuntimeProps { readonly plan: LayoutRuntimePlan; readonly onSubmit?: () => void }

/** The five binding kinds a block may carry (DES-0052 layout-ck-21 to layout-ck-25). */
export type LayoutBindingKind = 'record_field' | 'query' | 'measure' | 'template' | 'static'

/**
 * One authored binding: the kind, and the name it resolves by. A binding whose `name` is
 * empty is the `needs a binding` state — the block is preserved and rebound one at a time
 * rather than deleted (layout-auth-31).
 */
export interface LayoutAuthoringBinding {
  readonly kind: LayoutBindingKind
  readonly name: string
}

/**
 * What a capture block narrows (layout-ck-30): it may add a requirement and name registered
 * validation rules (layout-auth-21), never remove what Records declared.
 */
export interface LayoutAuthoringCapture {
  readonly required?: boolean
  readonly validationRules?: readonly string[]
  /** layout-auth-22: the prompt this surface shows for the field, in its own context only. */
  readonly promptOverride?: string
  /** layout-bound-3: the registered field control this capture field uses; absent is the runtime's choice. */
  readonly control?: string
}

/** A Rules named predicate's exact pin (rules-ck-22): nothing floats. */
export interface LayoutPredicatePin { readonly name: string; readonly version: string; readonly digest: string }

/**
 * layout-ck-29: a guard holds exactly one of a Rules expression, compiled at publish, or a named
 * predicate by exact pin, resolved from the definition's pinned closure. Both fail closed.
 */
export type LayoutAuthoringShowWhen =
  | { readonly expression: string; readonly predicate?: never }
  | { readonly predicate: LayoutPredicatePin; readonly expression?: never }

export interface LayoutAuthoringBlock {
  readonly id: string
  readonly kind: string
  readonly binding?: LayoutAuthoringBinding
  /** The block this one is authored inside; absent means the surface root. */
  readonly parentId?: string
  /** layout-auth-18: the block repeats its children once per row of its collection binding. */
  readonly repeating?: boolean
  /** The flow this block arranges its children in; admission requires one on a repeating block or any parent (T-724 ruling 41). */
  readonly container?: LayoutContainerFlow
  /** layout-auth-19: the declared Records relationship this block observes; only the key is stored. */
  readonly relatedRelationship?: string
  /** layout-ck-29, layout-auth-20: the block's guard, holding exactly one form; absent, the block always shows. */
  readonly showWhen?: LayoutAuthoringShowWhen
  /** The guided expression `showWhen` was lowered from; absent when the guard was written as raw text (T-724 ruling 39). */
  readonly showWhenGuide?: RuleDefinitionExpression
  /** The capture properties a capture block narrows with (layout-ck-30). */
  readonly capture?: LayoutAuthoringCapture
  /** layout-auth-33: the selection this block opens with. Authored; the live selection is never stored. */
  readonly defaultSelection?: string
  /** layout-auth-34: the ids of the other blocks on this surface this block's selection filters. */
  readonly filterTargets?: readonly string[]
  readonly intent?: LayoutIntent
  readonly zone?: string
  readonly width?: LayoutSizing
  readonly height?: LayoutSizing
  readonly justifySelf?: LayoutAlignment
  readonly alignSelf?: LayoutAlignment
  readonly staticRegion?: string
  readonly widgetId?: string
  readonly breakBefore?: boolean
  readonly avoidPageBreak?: boolean
}
export interface LayoutAuthoringPageRun { readonly id: string; readonly pageLayoutId: string; readonly pageMasterId: string }
export interface LayoutAuthoringDraft {
  readonly name: string
  readonly medium: LayoutMedium
  readonly defaultIntent?: LayoutIntent
  readonly containerFlow?: LayoutContainerFlow
  readonly collapseBelow?: LayoutContainerToken
  readonly density?: 'comfortable' | 'compact'
  readonly gap?: number
  readonly blocks: readonly LayoutAuthoringBlock[]
  readonly pageRuns?: readonly LayoutAuthoringPageRun[]
  /** layout-auth-35: the released surfaces a reader may drill through to from this one. */
  readonly drillThroughTargets?: readonly string[]
}
/** One name a binding picker can offer for its kind. */
export interface LayoutBindableName { readonly id: string; readonly label: string }

export interface LayoutAuthoringCatalogue {
  readonly blockKinds: readonly { readonly id: string; readonly label: string }[]
  readonly zones: readonly string[]
  /** The names offered per binding kind; `static` is authored on the block, so it is absent. */
  readonly bindables?: Partial<Record<Exclude<LayoutBindingKind, 'static'>, readonly LayoutBindableName[]>>
  readonly pageLayouts?: readonly { readonly id: string; readonly label: string }[]
  readonly pageMasters?: readonly { readonly id: string; readonly label: string }[]
  readonly staticRegions?: readonly string[]
  readonly helmWidgets?: readonly { readonly id: string; readonly label: string }[]
  /** The Records relationships declared on the surface's record type, by key (layout-auth-19). */
  readonly relationships?: readonly LayoutBindableName[]
  /** The record fields Records declares required; a capture block cannot drop them (layout-auth-21). */
  readonly requiredFields?: readonly string[]
  /** The registered validation rules a capture block may name (layout-auth-21, layout-bound-8). */
  readonly validationRules?: readonly LayoutBindableName[]
  /** The released surfaces a drill-through may name (layout-auth-35). */
  readonly drillTargets?: readonly LayoutBindableName[]
  /** The field controls the host registers for capture fields (layout-bound-3). */
  readonly fieldControls?: readonly LayoutBindableName[]
  /** The record fields whose value domain picks their editor; they take no authored control (layout-bound-10). */
  readonly valueDomainFields?: readonly string[]
  /** The references a show_when guard may read, offered by the guided expression editor (layout-auth-20). */
  readonly guardReferences?: readonly { readonly id: string; readonly label: string; readonly valueType: RuleDefinitionValueType }[]
  /** The named predicates a show_when guard may cite, each by its exact pin (layout-ck-29). */
  readonly predicates?: readonly { readonly label: string; readonly pin: LayoutPredicatePin }[]
}
export interface LayoutAuthoringEditorProps { readonly value: LayoutAuthoringDraft; readonly catalogue: LayoutAuthoringCatalogue; readonly onChange: (value: LayoutAuthoringDraft) => void }
